modules/bmc/bmc_api.rb

```diff
@@ -208,50 +217,80 @@ class Api < ::Sinatra::Base
       provider_type.downcase! if provider_type
+      if log_level = ::Proxy::BMC::Plugin.settings.provider_log_level
+        # using the log_level below will cause rubyipmi to create its own log, instead of using the proxy log
+        begin
+          log_level = ::Logger.const_get(log_level.upcase)
+        rescue
+          logger.warn("Incorrect bmc log level supplied #{log_level}, ignoring setting")
+          log_level = nil
+        end
+      end
+      # unless the provider is shell find a suitable provider
+      if provider_type != 'shell'
+        provider_type = find_ipmi_provider(provider_type)
+      end
 
       case provider_type
-      when /ipmi/
-        require 'bmc/ipmi'
+        when 'freeipmi', 'ipmitool'
 
-        raise "unauthorized" unless auth.provided?
-        raise "bad_authentication_request" unless auth.basic?
+          raise "unauthorized" unless auth.provided?
+          raise "bad_authentication_request" unless auth.basic?
 
-        username, password = auth.credentials
+          username, password = auth.credentials
 
-        # check to see if provider is given and no default provider is set, search for installed providers
-        unless Proxy::BMC::IPMI.installed?(provider_type)
-          # check if provider_type is a valid type
-          if Proxy::BMC::IPMI.providers.include?(provider_type)
-            log_halt 400, "#{provider_type} is not installed, please install a ipmi provider"
+          # setting the log level must come before IPMI initialization
+          # if the user supplies the provider_log_level all provider logs will be sent to
+          # the provider specific log instead of the smart proxy logs
+          # if the user doesn't supply the provider log level, all logs will be sent to the smart proxy logs
+          # this was configured to make easier specific log output for BMC without having to see other logs
+          if log_level
+            Proxy::BMC::IPMI.log_level = log_level
           else
-            log_halt 400, "Unrecognized or missing bmc provider type: #{provider_type}"
+            # this causes rubyipmi to use the supplied logger, most actions in rubyipmi only output during Logger::DEBUG
+            Proxy::BMC::IPMI.logger = logger
           end
-        end
 
-        # all the use of the http auth basic header to pass credentials
-        args = { :host     => params[:host], :username     => username,
-                 :password => password,      :bmc_provider => provider_type }
-        @bmc = Proxy::BMC::IPMI.new(args)
-      when "shell"
-        require 'bmc/shell'
-        @bmc = Proxy::BMC::Shell.new
-      else
-        log_halt 400, "Unrecognized or missing BMC type: #{provider_type}"
+          # all the use of the http auth basic header to pass credentials
+          args = { :host     => params[:host], :username     => username,
+                   :password => password,      :bmc_provider => provider_type }
```

Comment:
The idea here is to pass the credentials on the command line to execute ipmi commands and has nothing to do with authorization of using smart-proxy.  Its simply a tunnel to pass credentials through, since we are essentially remotely executing ipmi commands using Rubyipmi.  At this time Rubyipmi does not support certificates so neither does the bmc smart-proxy module.  I think it would be awesome to rely on a certificate but a lot more work would be needed to make this happen which is completely out of scope for this PR.  Furthermore, passing credentials is the easiest thing to do and using a ssl cert would probably be the next line of business for this module and Rubyipmi.